index.js

```diff
@@ -4,49 +4,113 @@ const app = express()
 const fs = require('fs');
 const { promisify } = require('util')
 const readFile = promisify(fs.readFile)
+const GPT_MODE = process.env.GPT_MODE
 
-app.use(express.json({extended: true, limit: '1mb'}))
+let file_context = "You are a helpful Twitch Chatbot."
+
+const messages = [
+  {role: "system", content: "You are a helpful Twitch Chatbot."}
+];
 
+console.log("GPT_MODE is " + GPT_MODE)
+console.log("History length is " + process.env.HISTORY_LENGTH)
+console.log("OpenAI API Key:" + process.env.OPENAI_API_KEY)
+
+app.use(express.json({extended: true, limit: '1mb'}))
 
 app.all('/', (req, res) => {
     console.log("Just got a request!")
     res.send('Yo!')
 })
 
-let file_context = ""
-fs.readFile("./file_context.txt", 'utf8', function(err, data) {
-  if (err) throw err;
-  console.log(file_context);
-  file_context = data;
-});
+if (process.env.GPT_MODE === "CHAT"){
+
+  fs.readFile("./file_context.txt", 'utf8', function(err, data) {
+    if (err) throw err;
+    console.log("Reading context file and adding it as system level message for the agent.")
+    messages[0].content = data;
+  });
+
+} else {
+
+  fs.readFile("./file_context.txt", 'utf8', function(err, data) {
+    if (err) throw err;
+    console.log("Reading context file and adding it in front of user prompts:")
+    file_context = data;
+    console.log(file_context);
+  });
+
+}
 
 app.get('/gpt/:text', async (req, res) => {
+    
+    //The agent should recieve Username:Message in the text to identify conversations with different users in his history. 
+    
     const text = req.params.text
     const { Configuration, OpenAIApi } = require("openai");
 
-    console.log(process.env.OPENAI_API_KEY)
     const configuration = new Configuration({
       apiKey: process.env.OPENAI_API_KEY,
     });
-    const openai = new OpenAIApi(configuration);
 
-    const prompt = file_context + "\n\nQ:" + text + "\nA:";
-    console.log(prompt);
+    const openai = new OpenAIApi(configuration);      
     
-    const response = await openai.createCompletion({
-      model: "text-davinci-003",
-      prompt: prompt,
-      temperature: 0.5,
-      max_tokens: 128,
-      top_p: 1,
-      frequency_penalty: 0,
-      presence_penalty: 0,
-    });
-    if (response.data.choices) {
-        res.send(response.data.choices[0].text)
-    } else {
+    if (GPT_MODE === "CHAT"){
+      //CHAT MODE EXECUTION
+
+      //Add user message to  messages
+      messages.push({role: "user", content: text})
```

Comment:
I was wondering, what would be the pros and cons of having a specific user history vs chat history?

The username could be easly sent. 

Let me know what you think about it.

Reply:
I thought about saving the last few messages of each user to the db and only sending the last messages of that user to GPT but that removes the ability for the bot to understand the ongoing discussion. So Viewer A could not amend Viewer B.

In my extensive testing in a 50+ User Stream Chat the bot understands different users in his history and if somebody writes "could you write what you told user X in a bit more funny way" the bot will respond. 

The way it is implemented right now lets the bot feel more like a person that is also part of the community instead of a personal chatbot.

Reply:
Interesting. 

Let's keep this way. Thank you so much for the answer and the insights :)